src/encoder.rs

```diff
@@ -807,140 +808,580 @@ impl<W: Write> Drop for Writer<W> {
+            fctl.y_offset = 0;
+            Ok(())
+        } else {
+            Err(EncodingError::Format(FormatErrorKind::NotAnimated.into()))
+        }
+    }
+
+    /// Set the blend operation for the following frames.
+    ///
+    /// The blend operation specifies whether the frame is to be alpha blended
+    /// into the current output buffer content, or whether it should completely
+    /// replace its region in the output buffer.
+    ///
+    /// See the [`BlendOp`] documentaion for the possible values and their effects.
+    ///
+    /// *Note that for the first frame the two blend modes are functionally
+    /// equivalent due to the clearing of the output buffer at the beginning
+    /// of each play.*
+    ///
+    /// This method will return an error if the image is not animated.
+    ///
+    /// [`BlendOP`]: enum.BlendOp.html
+    pub fn set_blend_op(&mut self, op: BlendOp) -> Result<()> {
+        if let Some(ref mut fctl) = self.fctl {
+            fctl.blend_op = op;
+            Ok(())
+        } else {
+            Err(EncodingError::Format(FormatErrorKind::NotAnimated.into()))
+        }
+    }
+
+    /// Set the dispose operation for the following frames.
+    ///
+    /// The dispose operation specifies how the output buffer should be changed
+    /// at the end of the delay (before rendering the next frame)
+    ///
+    /// See the [`DisposeOp`] documentaion for the possible values and their effects.
+    ///
+    /// *Note that if the first frame uses [`DisposeOp::Previous`]
+    /// it will be treated as [`DisposeOp::Background`].*
+    ///
+    /// This method will return an error if the image is not animated.
+    ///
+    /// [`DisposeOp`]: ../common/enum.BlendOp.html
+    /// [`DisposeOp::Previous`]: ../common/enum.BlendOp.html#variant.Previous
+    /// [`DisposeOp::Background`]: ../common/enum.BlendOp.html#variant.Background
+    pub fn set_dispose_op(&mut self, op: DisposeOp) -> Result<()> {
+        if let Some(ref mut fctl) = self.fctl {
+            fctl.dispose_op = op;
+            Ok(())
+        } else {
+            Err(EncodingError::Format(FormatErrorKind::NotAnimated.into()))
         }
     }
 
     pub fn finish(mut self) -> Result<()> {
+        if !self.end {
+            let err = FormatErrorKind::MissingFrames.into();
+            return Err(EncodingError::Format(err));
+        } else if self.to_write > 0 {
+            let err = FormatErrorKind::MissingData(self.to_write).into();
+            return Err(EncodingError::Format(err));
+        }
         // TODO: call `writer.finish` somehow?
         self.flush()?;
         Ok(())
     }
+
+    /// Flushes the buffered chunk, checks if it was the last frame,
+    /// writes the next frame header and gets the next frame scanline size
+    /// and image size.
+    fn new_frame(&mut self) -> Result<()> {
+        let mut wrt = match self.writer.take() {
+            Wrapper::Chunk(wrt) => wrt,
+            _ => unreachable!(),
```

Comment:
Can you add a short explanation of why this can't be reached? It would appear to me that an error in `Writer::write` for example leaves it in the state `Wrapper::None` but the instance still exists.

Reply:
Yes this should not `take()` but borrow instead, it should `take()` only right before constructing the `ZlibEncoder`